File: update_project_statuses.py
#!/usr/bin/env python3
"""
Script to update GitHub project item statuses.
You need to have a GitHub token with project write access.
Set it as an environment variable GITHUB_TOKEN before running.
"""

import os
import json
import subprocess
import time

# Project information
PROJECT_ID = "PVT_kwHODHqkes4A4KTr"  # Turbo Modern Starter Implementation
STATUS_FIELD_ID = "PVTSSF_lAHODHqkes4A4KTrzgtLOB4"  # Status field

# Status options
STATUS_OPTIONS = {
    "todo": "f75ad846",       # Todo
    "in_progress": "47fc9ee4", # In Progress
    "review": "a352001e",     # Pending Review
    "done": "98236657",       # Done
    "on_hold": "0888ff68"     # On Hold
}

# Item IDs and desired statuses
ITEMS_TO_UPDATE = {
    # Issue #61 (UI component library)
    "PVTI_lAHODHqkes4A4KTrzgZ_xhs": "in_progress",
    
    # Issue #62 (i18n support)
    "PVTI_lAHODHqkes4A4KTrzgZ_xh4": "todo",
    
    # Issue #63 (Data sync)
    "PVTI_lAHODHqkes4A4KTrzgZ_xh8": "todo",
    
    # PR #64 (UI component library)
    "PVTI_lAHODHqkes4A4KTrzgZ_xvM": "review",
    
    # Issue #65 (GitHub workflow)
    "PVTI_lAHODHqkes4A4KTrzgZ_x1M": "in_progress",
    
    # PR #66 (GitHub workflow guidelines)
    "PVTI_lAHODHqkes4A4KTrzgZ_x1o": "review",
    
    # PR #67 (GitHub project automation)
    "PVTI_lAHODHqkes4A4KTrzgZ_yQc": "review"
}

def update_status(item_id, status_option):
    """Update the status of a project item."""
    option_id = STATUS_OPTIONS[status_option]
    print(f"Updating item {item_id} to status '{status_option}' (option_id: {option_id})...")
    
    # GraphQL mutation
    mutation = f"""
    mutation {{
      updateProjectV2ItemFieldValue(
        input: {{
          projectId: "{PROJECT_ID}"
          itemId: "{item_id}"
          fieldId: "{STATUS_FIELD_ID}"
          value: {{
            singleSelectOptionId: "{option_id}"
          }}
        }}
      ) {{
        projectV2Item {{
          id
        }}
      }}
    }}
    """
    
    # Execute GraphQL query
    command = ["gh", "api", "graphql", "-f", f"query={mutation}"]
    result = subprocess.run(command, capture_output=True, text=True)
    
    if result.returncode != 0:
        print(f"Error: {result.stderr}")
        print(f"Response: {result.stdout}")
        return False
    
    print("Status updated successfully")
    return True

def main():
    """Main function to update all project item statuses."""
    if "GITHUB_TOKEN" not in os.environ:
        print("Error: GITHUB_TOKEN environment variable not set")
        print("Please set it with: export GITHUB_TOKEN=your_token")
        return
    
    success_count = 0
    failure_count = 0
    
    for item_id, status in ITEMS_TO_UPDATE.items():
        if update_status(item_id, status):
            success_count += 1
        else:
            failure_count += 1
        
        # Sleep briefly to avoid rate limiting
        time.sleep(1)
    
    print(f"\nSummary: {success_count} items updated successfully, {failure_count} failures")

if __name__ == "__main__":
    main()

File: .claude/scripts/update-project-statuses.js
#!/usr/bin/env node

/**
 * update-project-statuses.js
 * 
 * Automatically updates GitHub project item statuses based on predefined mappings.
 * This script uses the GitHub GraphQL API to update project items.
 * 
 * Environment variables:
 * - GITHUB_TOKEN: Personal access token with project access
 * - PROJECT_ID: ID of the GitHub project
 * - ORGANIZATION: GitHub organization name
 * - REPOSITORY: GitHub repository name
 */

const { graphql } = require('@octokit/graphql');
require('dotenv').config();

// Configuration
const PROJECT_ID = process.env.PROJECT_ID || 'PVT_kwHODHqkes4A4KTr';
const GITHUB_TOKEN = process.env.GITHUB_TOKEN;
const OWNER = process.env.ORGANIZATION || 'CodingButterBot';
const REPOSITORY = process.env.REPOSITORY || 'turbo-modern-starter';

// Status field ID - will be populated from the API
let STATUS_FIELD_ID = 'PVTSSF_lAHODHqkes4A4KTrzgtLOB4';

// Status option IDs - will be populated from the API
const STATUS_OPTIONS = {
  TODO: 'f75ad846',      // Todo status ID
  IN_PROGRESS: '47fc9ee4', // In Progress status ID
  REVIEW: 'a352001e',    // Pending Review status ID
  DONE: '98236657'       // Done status ID
};

// GraphQL client with authentication
const graphqlWithAuth = graphql.defaults({
  headers: {
    authorization: `token ${GITHUB_TOKEN}`,
  },
});

/**
 * Gets project field information to find status field and options
 */
async function getProjectFields() {
  try {
    const { user } = await graphqlWithAuth(`
      query {
        user(login: "${OWNER}") {
          projectV2(number: 4) {
            id
            fields(first: 20) {
              nodes {
                ... on ProjectV2SingleSelectField {
                  id
                  name
                  options {
                    id
                    name
                  }
                }
              }
            }
          }
        }
      }
    `);

    return user.projectV2.fields.nodes;
  } catch (error) {
    console.error('Error fetching project fields:', error.message);
    throw error;
  }
}

/**
 * Gets items from a GitHub project
 */
async function getProjectItems() {
  try {
    const { user } = await graphqlWithAuth(`
      query {
        user(login: "${OWNER}") {
          projectV2(number: 4) {
            id
            items(first: 100) {
              nodes {
                id
                content {
                  ... on Issue {
                    title
                    number
                    state
                    closed
                    repository {
                      name
                    }
                  }
                  ... on PullRequest {
                    title
                    number
                    state
                    isDraft
                    closed
                    merged
                    repository {
                      name
                    }
                  }
                }
                fieldValues(first: 8) {
                  nodes {
                    ... on ProjectV2ItemFieldSingleSelectValue {
                      name
                      field {
                        ... on ProjectV2SingleSelectField {
                          name
                          id
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    `);

    return user.projectV2.items.nodes;
  } catch (error) {
    console.error('Error fetching project items:', error.message);
    throw error;
  }
}

/**
 * Updates the status of a project item
 */
async function updateItemStatus(itemId, optionId) {
  try {
    const result = await graphqlWithAuth(`
      mutation {
        updateProjectV2ItemFieldValue(
          input: {
            projectId: "${PROJECT_ID}"
            itemId: "${itemId}"
            fieldId: "${STATUS_FIELD_ID}"
            value: { 
              singleSelectOptionId: "${optionId}"
            }
          }
        ) {
          projectV2Item {
            id
          }
        }
      }
    `);

    return result;
  } catch (error) {
    console.error(`Error updating item ${itemId}:`, error.message);
    throw error;
  }
}

/**
 * Determines the appropriate status for an item based on its content
 */
function determineItemStatus(item) {
  if (!item.content) {
    return STATUS_OPTIONS.TODO;
  }

  const content = item.content;
  
  // For pull requests
  if (content.__typename === 'PullRequest') {
    if (content.merged) {
      return STATUS_OPTIONS.DONE;
    } else if (content.closed) {
      return STATUS_OPTIONS.DONE; // Closed PRs should also be Done
    } else if (content.isDraft) {
      return STATUS_OPTIONS.IN_PROGRESS;
    } else {
      return STATUS_OPTIONS.REVIEW; // Open, non-draft PRs are in review
    }
  } 
  // For issues
  else if (content.__typename === 'Issue') {
    if (content.closed) {
      return STATUS_OPTIONS.DONE;
    }
    
    // Check for any current status to preserve it if it's more advanced
    const currentStatus = getCurrentFieldValue(item, 'Status');
    
    if (currentStatus) {
      // If it's already in review or in progress, preserve that status
      if (currentStatus.toLowerCase() === 'in progress') {
        return STATUS_OPTIONS.IN_PROGRESS;
      } else if (currentStatus.toLowerCase().includes('review')) {
        return STATUS_OPTIONS.REVIEW;
      }
    }
    
    // By default, open issues are Todo
    return STATUS_OPTIONS.TODO;
  }
  
  // Default case
  return STATUS_OPTIONS.TODO;
}

/**
 * Gets the current value of a field for an item
 */
function getCurrentFieldValue(item, fieldName) {
  const fieldValue = item.fieldValues.nodes.find(node => 
    node.field && node.field.name === fieldName
  );
  return fieldValue ? fieldValue.name : null;
}

/**
 * Main function
 */
async function main() {
  try {
    console.log('Fetching project fields...');
    const fields = await getProjectFields();
    
    // Find Status field and map options
    const statusField = fields.find(field => field.name === 'Status');
    if (!statusField) {
      throw new Error('Status field not found in the project');
    }
    
    STATUS_FIELD_ID = statusField.id;
    console.log('Status field found:', STATUS_FIELD_ID);
    
    // Map status options
    for (const option of statusField.options) {
      const name = option.name.toLowerCase();
      if (name === 'todo') {
        STATUS_OPTIONS.TODO = option.id;
      } else if (name === 'in progress') {
        STATUS_OPTIONS.IN_PROGRESS = option.id;
      } else if (name.includes('review') || name === 'pending review') {
        STATUS_OPTIONS.REVIEW = option.id;
      } else if (name === 'done') {
        STATUS_OPTIONS.DONE = option.id;
      }
    }
    
    console.log('Status options mapped:', STATUS_OPTIONS);
    
    // Get all project items
    console.log('Fetching project items...');
    const items = await getProjectItems();
    
    console.log(`Found ${items.length} items in the project`);
    
    // Process each item
    let updatedCount = 0;
    
    for (const item of items) {
      if (!item.content || item.content.repository.name !== REPOSITORY) {
        continue; // Skip items not from our repository
      }
      
      const itemId = item.id;
      const content = item.content;
      const contentTitle = content.title;
      const contentNumber = content.number;
      const contentType = content.__typename || (content.repository ? 'Issue' : 'Unknown');
      
      console.log(`Processing ${contentType} #${contentNumber}: ${contentTitle}`);
      
      // Get current status
      const currentStatus = getCurrentFieldValue(item, 'Status');
      console.log(`  Current status: ${currentStatus || 'Not set'}`);
      
      // Determine appropriate status
      const newStatusId = determineItemStatus(item);
      
      // Convert IDs back to names for logging
      const newStatusName = Object.keys(STATUS_OPTIONS).find(key => 
        STATUS_OPTIONS[key] === newStatusId
      );
      
      console.log(`  Appropriate status: ${newStatusName}`);
      
      // Check if status needs to be updated
      const currentStatusId = statusField.options.find(o => 
        o.name === currentStatus
      )?.id;
      
      if (currentStatusId !== newStatusId) {
        console.log(`  Updating status to ${newStatusName}...`);
        await updateItemStatus(itemId, newStatusId);
        updatedCount++;
      } else {
        console.log(`  Status already correct`);
      }
    }
    
    console.log(`Updated ${updatedCount} items`);
  } catch (error) {
    console.error('Error in main function:', error);
    process.exit(1);
  }
}

main();